Set correct output dtype for dequantize op during convert_pt2e in decomposed mode

kausikmaiti opened the pull request:
Earlier the signature of dequantize ops for decomposed quantized Tensor was changed for wider use-cases where the output dtype can be different from torch.float and needs to be passed during dequantization.
Please refer: https://github.com/pytorch/pytorch/pull/121450

However, setting of correct output dtype for dequantize ops was still missing in convert_pt2e flow.

This change enables the users to use PT2E quantization flow with non torch.float unquantized dtype, such as torch.bfloat16.

kausikmaiti committed:
Set correct output dtype for dequantize op
during convert_pt2e in decomposed mode

kausikmaiti commented:
Hi @jerryzh168, Can you please review this change? Thank you.

kausikmaiti commented:
Hi, I'm looking forward to the review comments on this change. Kindly help in this regard. Thank you.

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1233,15 +1234,51 @@ def forward(self, x):
+        def verify_quant_dequant_iotypes(m, activation_dtype, activation_quant_dtype):
+            quant_node = None
+            for node in m.graph.nodes:
+                if (
```
nit: I think it might be more robust and easier to understand to
1. check for dequantize node
2. traverse back to find the corresponding quantize node
    (quant_node = dequant_node.args[0])
4. do all the checks for quant and dequant node

jerryzh168 reviewed (approved):
looks good, thanks for the fix! had a comment to refactor test a bit

jerryzh168 commented:
also need to fix CI, there are a lot of breaking tests

jgong5 reviewed (changes requested):
Please fix UTs and tag me for review again.

kausikmaiti committed:
Incorporated review comments and added fix for CI test failures

Signed-off-by: kausik <kmaiti@habana.ai>

kausikmaiti committed:
Deleted unused code

Signed-off-by: kausik <kmaiti@habana.ai>

kausikmaiti commented:
Hi, I've incorporated review comments. Also, I added fix for the test failures.
@jgong5, @jerryzh168, Kindly review and let me know whether this revised version is fine.
Thank you.

jgong5 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1243,41 +1243,36 @@ def forward(self, x):
-        # Assumption is good enough for the simple model used for testing at present
-        def verify_quant_dequant_iotypes(m, activation_dtype, activation_quant_dtype):
-            quant_node = None
+        def verify_quant_dequant_iotypes_new(m):
```
hmm, why changing the name?

kausikmaiti replied:
Sorry for this unintended change. Let me correct.

kausikmaiti committed:
Corrected function name

Signed-off-by: kausik <kmaiti@habana.ai>

jgong5 commented:
@pytorchbot merge

jgong5 commented:
@pytorchbot rebase

jgong5 commented:
@pytorchbot merge

kausikmaiti commented:
@pytorchbot rebase

kausikmaiti commented:
@jgong5, @jerryzh168, Can you please help to merge this change? Thank you.

jgong5 commented:
@pytorchbot merge
